src/Korobi/WebBundle/Util/GitInfo.php

```diff
@@ -13,45 +13,130 @@ class GitInfo {
      */
-    public function __construct($environment, $rootDir) {
-        $this->updateData($rootDir, $environment);
+    public function __construct($environment, $appDir) {
+        $this->updateData(dirname($appDir), $environment);
     }
 
     /**
      * Force update data.
      *
-     * @param string $environment
+     * @param string $rootDir Absolute path to root directory.
+     * @param string $environment Current environment's name.
      */
-    public function updateData($rootDir, $environment = 'dev') {
-        // Fast information retrieval for online environments
-        if (in_array($environment, self::ONLINE_ENVS)) {
-            if ($environment == 'prod') {
-                $this->branch = 'www1-stable';
+    public function updateData($rootDir, $environment) {
+        // This process is not simple! Please be careful when maintaining this code.
+        // Please refer to the flowchart in the PR for further information on how this works.
+        // https://github.com/korobi/Web/pull/211
+
+        // Initialize the stack of functions to use
+        $process = [];
+        // The first step is to ensure we have a valid .git folder
+        $process[] = "checkIsGitRepo";
+        // Next, we'll make sure there's a .git/HEAD file and process it
+        $process[] = "checkForGitHeadFile";
+        // These are now fallbacks, for if the .git/HEAD approach failed
+        // Here, we look for the packed-refs file and parse it
+        $process[] = "checkForPackedRefs";
+        // Finally, if all else fails we use bendem's shell command approach
+        // this is a bit slower but hopefully is more reliable.
+        $process[] = "useFallbackCommand";
+
+        $gitDir = $rootDir . DIRECTORY_SEPARATOR . ".git" . DIRECTORY_SEPARATOR;
+        $firstItem = array_shift($process);
+        /** @var $firstItem Callable */
+        $this->$firstItem($gitDir, $process);
+
+        // Handle capifony
+        $this->postCheckForCapifonyFiles($gitDir, $environment);
+    }
+
+    /*
+     * Begin chain-of-responsibility function definitions
+     * --------------------------------------------------
+     */
+
+    private function checkIsGitRepo($gitDir, $stack) {
+        if (file_exists($gitDir)) {
+            /** @var Callable $next */
+            $next = array_shift($stack);
+            $this->$next($gitDir, $stack);
+        } else {
+            $this->setBranchAndHash("unknown", "unknown");
+        }
+    }
+
+    private function checkForGitHeadFile($gitDir, $stack) {
+        if (file_exists($gitDir . "HEAD")) {
+            $headContents = fgets(fopen($gitDir . "HEAD", 'r')); // resource acquisition is initialization
+            $this->handleGitHeadFile($gitDir, $headContents, $stack);
+        } else {
+            $this->setBranchAndHash("unknown", "unknown");
+        }
+    }
+
+    private function handleGitHeadFile($gitDir, $headContents, $stack) {
+        if (strpos($headContents, "ref: ") === -1) {
```

Comment:
Changing `file_exists` to `is_readable` should cover all the bases when fgets can return false. 

The only other scenario I can think of is when HEAD is empty, not aware of any case when this would happen with a real git repository.